### PR TITLE
nat: Do not inc error metric from CT GC if related NAT entry not found

### LINKED ISSUE #11485: nat: Do not inc error metric from CT GC if related NAT entry not found
Currently, when removing an entry from a BPF map via `(*pkg/bpf.Map) Delete(...)`, a related error metric is incremented if the entry does not exist (`ENOENT`). This leads to many false positive map entry deletion errors when CT GC tries to remove a related NAT entry without checking whether it exists (e.g. `snat_v4_external_delete` reported by @tgraf).

To fix this we could extend `(*pkg/bpf.Map) Delete(...)` by adding a param which controls whether `ENOENT` is expected, and if it's so, then the error metric won't be incremented if the entry is not found.

### PR DESCRIPTION
**Pull request by soumynathan**
This PR adds in an additional param to bpf.Map Delete() so that the
error metric count increment can be controlled if the entry not found.

Fixes: #11485
Signed-off-by: Swaminathan Vasudevan <svasudevan@suse.com>

### CODE REVIEW

**Comment by soumynathan** [comment]
Work in Progress.

**Comment by mrostecki** [comment]
Since it's a progress, I marked is as a draft.

**Review by aanm — CHANGES_REQUESTED** [review]
It seems to me that `ignoreMetricCountifNotAvailable` is always false on all cases?

**Comment by soumynathan** [comment]
> It seems to me that `ignoreMetricCountifNotAvailable` is always false on all cases?

Yes you are right, this is to provide an option to configure the Metric. For now I have not set it to 'true' as @tgraf requested for the CT maps. I just wanted to confirm if we are going in the right direction and then we can enable the CT maps that has this issue.

File: pkg/bpf/map_linux.go
Comment (brb) [inline pkg/bpf/map_linux.go]: Instead of changing the `Delete` method, have you considered creating a new method `DeleteWithIgnoreFoo` which would be then used by the CT? This approach would not require changing all the existing invocations of `Delete`?

Comment (soumynathan) [reply pkg/bpf/map_linux.go]: @brb good point, let me consider this. My initial understanding is to add an argument to the Delete. But I did realize that I am touching too many files to make this change and it is worth while to just add a new function.

File: pkg/maps/ctmap/ctmap.go
Comment (brb) [inline pkg/maps/ctmap/ctmap.go]: You probably want to use the method at the line 299 instead.

File: pkg/maps/ctmap/ctmap.go
Comment (brb) [inline pkg/maps/ctmap/ctmap.go]: Same here (line 379).

File: pkg/bpf/map_linux.go
Comment (brb) [inline pkg/bpf/map_linux.go]: I'd include an explanation for when it can happen (when element doesn't exist).

File: pkg/bpf/map_linux.go
Comment (brb) [inline pkg/bpf/map_linux.go]: IMHO, this method name reads as "this method deletes an ignore metric error counter".

**Review by brb — CHANGES_REQUESTED** [review]

File: pkg/bpf/map_linux.go
Comment (joestringer) [reply pkg/bpf/map_linux.go]: Should we call it something like `SilentDelete()` or something because it silences metrics which could be used to figure out what it's doing?

File: pkg/maps/ctmap/ctmap.go
Comment (brb) [inline pkg/maps/ctmap/ctmap.go]: Why did you remove the CT entry removal?

**Review by brb — CHANGES_REQUESTED** [review]

File: pkg/maps/ctmap/ctmap.go
Comment (soumynathan) [reply pkg/maps/ctmap/ctmap.go]: @brb I think I am confused here. I thought you want me to replace the 'DeleteIgnoreMetricErrorCount' function with 'DeleteMapping' which is at line 299.
Can you clarify here.

File: pkg/maps/ctmap/ctmap.go
Comment (soumynathan) [reply pkg/maps/ctmap/ctmap.go]: Sorry the comment was not clear. I should have asked for clarity.

**Comment by aanm** [comment]
@soumynathan are you waiting for review or this is still WIP? Either case please re-ask for a review from Martynas (brb).

**Comment by soumynathan** [comment]
> @soumynathan are you waiting for review or this is still WIP? Either case please re-ask for a review from Martynas (brb) once the PR is done.
I am still waiting for review comments from @brb. This is not WIP.

**Comment by aanm** [comment]
@soumynathan the failure seems to be related with the changes:
```
make[1]: Entering directory '/home/travis/gopath/src/github.com/cilium/cilium/proxylib'
CGO_ENABLED=1 go build  -mod=vendor -ldflags ' -X "github.com/cilium/cilium/pkg/version.Version=1.8.90 b5081a9 2020-06-12T06:31:25+00:00 go version go1.14.4 linux/amd64" -s -w -X "github.com/cilium/cilium/pkg/envoy.RequiredEnvoyVersionSHA=a8f292139e923b205525feb2c8a4377005904776" -X "github.com/cilium/cilium/pkg/datapath/loader.DatapathSHA=826e122bb500b65d5fa64156f186bf7742890359" -extldflags -Wl,-soname,libcilium.so.1'  -o libcilium.so.1 -buildmode=c-shared
/home/travis/gopath/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:269:2: ineffectual assignment to value2
Makefile:416: recipe for target 'ineffassign' failed
make: *** [ineffassign] Error 1
make: *** Waiting for unfinished jobs....
ln -sf libcilium.so.1 libcilium.so || cp libcilium.so.1 libcilium.so
make[1]: Leaving directory '/home/travis/gopath/src/github.com/cilium/cilium/proxylib'
```

**Comment by soumynathan** [comment]
> @soumynathan the failure seems to be related with the changes:
> 
> ```
> make[1]: Entering directory '/home/travis/gopath/src/github.com/cilium/cilium/proxylib'
> CGO_ENABLED=1 go build  -mod=vendor -ldflags ' -X "github.com/cilium/cilium/pkg/version.Version=1.8.90 b5081a9 2020-06-12T06:31:25+00:00 go version go1.14.4 linux/amd64" -s -w -X "github.com/cilium/cilium/pkg/envoy.RequiredEnvoyVersionSHA=a8f292139e923b205525feb2c8a4377005904776" -X "github.com/cilium/cilium/pkg/datapath/loader.DatapathSHA=826e122bb500b65d5fa64156f186bf7742890359" -extldflags -Wl,-soname,libcilium.so.1'  -o libcilium.so.1 -buildmode=c-shared
> /home/travis/gopath/src/github.com/cilium/cilium/pkg/bpf/map_linux_test.go:269:2: ineffectual assignment to value2
> Makefile:416: recipe for target 'ineffassign' failed
> make: *** [ineffassign] Error 1
> make: *** Waiting for unfinished jobs....
> ln -sf libcilium.so.1 libcilium.so || cp libcilium.so.1 libcilium.so
> make[1]: Leaving directory '/home/travis/gopath/src/github.com/cilium/cilium/proxylib'
> ```

Fixed.

File: pkg/bpf/map_linux.go
Comment (qmonnet) [inline pkg/bpf/map_linux.go:910]: This `if` condition seems to be the only difference with the original `Delete()` method. It makes little sense to have this chunk at all, since each time you call `DeleteButIgnoreMetricErrorCountInc()` you pass `true` as second argument.

As I understand the initial idea was to add an argument to `Delete()` itself, but that would involve changing the numerous calls to the method, so it's simpler to add a new function. But you can probably base `Delete()` on the new function instead, to avoid having too much duplicated code:

``` go
func (m *Map) DeleteButIgnoreMetricErrorCountInc(key MapKey, ignoreMetricErrorCount bool) error {

	[...]

	if option.Config.MetricsConfig.BPFMapOps && !ignoreMetricErrorCount {
		metrics.BPFMapOps.WithLabelValues(m.commonName(), metricOpDelete, metrics.Errno2Outcome(errno)).Inc()
	}

	[...]

}

func (m *Map) Delete(key MapKey) error {
	return DeleteButIgnoreMetricErrorCountInc(key, false)
}
```

In that case the boolean argument makes sense, and we avoid replicating the code.

Another note: as I understand, in the case we want to skip the metric update, we only want to skip it if the error is `ENOENT`, to avoid silencing errors other than the key not being found. So you might be missing a check here.

Please make sure to address the other suggestions from older comments (e.g. on function name).

**Review by joestringer — CHANGES_REQUESTED** [review]
I did not review but scanning through the page I see several unresolved comments.

Requesting changes, please re-request review when they have been addressed.

File: pkg/bpf/map_linux.go
Comment (joestringer) [reply pkg/bpf/map_linux.go:910]: @qmonnet I think this makes sense. FYI this looks like a "request changes" type feedback, but I see the "eye" icon from a review that selected to "Submit general feedback without explicit approval." Typically we try to avoid using this option to make it clear when changes should be addressed prior to merging, otherwise contributors do not get a clear signal whether action is required or not, and janitors do not get a clear signal about whether it is OK to merge. I've replied below requesting changes partially based on this and partially on the other feedback above.

Comment (joestringer) [reply pkg/bpf/map_linux.go:910]: (If this is just Github UI mismatch or my misintepretation, feel free to disregard my comment but just trying to bring some clarification to the state of this PR and suggestions to improve the review process for both contributors and maintainers :-) )

Comment (qmonnet) [reply pkg/bpf/map_linux.go:910]: @joestringer No problem and agreed, I can't remember why I went for “comment” on this particular review. Likely I went too fast :/. Thanks for the notice :)

**Comment by brb** [comment]
@soumynathan Are you still working on this?

**Comment by brb** [comment]
> @soumynathan Are you still working on this?

Bump.

**Comment by joestringer** [comment]
@qmonnet  / @brb I don't think we need to keep reopening and bumping the PR to keep it live if the author is not following up. It would be awesome if @soumynathan could pick this up and drive it forward, but if it's not a priority for him then I don't think it's going to make any difference to have us periodically bumping like this.

If someone is worried about making sure these changes make it in, then it'd be reasonable to pick up these changes where @soumynathan left off and either reopen the PR or repost for further review.

**Pull request closed without merging**
